sys/dev/usb/usb_generic.c

```diff
@@ -1223,54 +1203,175 @@ ugen_fs_copy_in(struct usb_fifo *f, uint8_t ep_index)
 			}
 
 			/* move data */
-			error = copyin(uaddr, kaddr, length);
+			error = copyin(fs_ep.ppBuffer[n], kaddr, length);
 			if (error) {
 				break;
 			}
 		}
 		offset += length;
 	}
+err:
+	free(fs_ep.ppBuffer, M_USB);
 	return (error);
 
 complete:
+	free(fs_ep.ppBuffer, M_USB);
 	mtx_lock(f->priv_mtx);
 	ugen_fs_set_complete(f, ep_index);
 	mtx_unlock(f->priv_mtx);
 	return (0);
 }
 
+static struct usb_fs_endpoint *
+ugen_fs_ep_uptr(struct usb_fifo *f, uint8_t ep_index)
+{
+	return ((struct usb_fs_endpoint *)
+	    ((char *)f->fs_ep_ptr + (ep_index * f->fs_ep_sz)));
+}
+
+static int
+ugen_fs_copyin(struct usb_fifo *f, uint8_t ep_index,
+    struct usb_fs_endpoint* fs_ep, bool copy_arrays)
+{
+	void *uptr;
+#ifdef COMPAT_FREEBSD32
+	struct usb_fs_endpoint32 fs_ep32;
+	uint32_t *uptr32;
+#endif
+	int error;
+
+	switch(f->fs_ep_sz) {
+	case sizeof(struct usb_fs_endpoint):
+		error = copyin(ugen_fs_ep_uptr(f, ep_index), fs_ep,
+		    f->fs_ep_sz);
+		uptr = fs_ep->ppBuffer;
+		if (error != 0)
+			return (error);
+		break;
+
+#ifdef COMPAT_FREEBSD32
+	case sizeof(struct usb_fs_endpoint32):
+		error = copyin(ugen_fs_ep_uptr(f, ep_index), &fs_ep32,
+		    f->fs_ep_sz);
+		if (error != 0)
+			return (error);
+		uptr32 = PTRIN(fs_ep32.ppBuffer);
+		PTRIN_CP(fs_ep32, *fs_ep, pLength);
+		CP(fs_ep32, *fs_ep, nFrames);
+		CP(fs_ep32, *fs_ep, aFrames);
+		CP(fs_ep32, *fs_ep, flags);
+		CP(fs_ep32, *fs_ep, timeout);
+		CP(fs_ep32, *fs_ep, isoc_time_complete);
+		CP(fs_ep32, *fs_ep, status);
+		break;
+#endif
+	default:
+		panic("%s: called with unknown size %#zx", __func__, f->fs_ep_sz);
+	}
+
+	if (!copy_arrays || fs_ep->nFrames == 0) {
+		fs_ep->ppBuffer = NULL;
+		return (0);
+	}
+
+	/*
+	 * Avoid userspace DoS.  16 is selected because it's the
+	 * largest number any in-tree driver uses (most use 0-2).
+	 * It might be that MAXPHYS/sizeof(void *) would be better.
+	 */
+	if (fs_ep->nFrames > 16)
+		return (EINVAL);
+
+	switch(f->fs_ep_sz) {
+	case sizeof(struct usb_fs_endpoint):
+		fs_ep->ppBuffer = malloc(sizeof(*fs_ep->ppBuffer) *
```

Comment:
The maximum number of frames / buffers is given when USB_FS_OPEN is called.
Please allocate this buffer there. Yes, it will be shared between threads, but only one thread should submit transfers at a time.

Or have a mixed stack / malloc usage, that when nFrames is small use stack, else malloc.

This function is in the fast path and can be called thousands of times per second and I don't want to see malloc there!
```
                        /* update number of frames */
                        u.popen->max_frames =
                            f->fs_xfer[u.popen->ep_index]->nframes;

```